Testing.Rmd

```diff
@@ -275,140 +382,640 @@ All expectations have a similar structure:
 
-*This passage is still being revised for the 2nd edition the R Packages book.
-In the meantime, please note that:
-in the third edition of testthat, `expect_equal()` and `expect_identical()` are both powered by the [waldo package]().
-In the past / in the second edition of testthat, `expect_equal()` wraps `all.equal()` and `expect_identical()` wraps `identical()`.*
-
-*   There are two basic ways to test for equality: `expect_equal()`, 
-    and `expect_identical()`. `expect_equal()` is the most commonly used: it 
-    checks for equality within a numerical tolerance:
+* There are two basic ways to test for equality: `expect_equal()`, and
+  `expect_identical()`.
+  `expect_equal()` is the most commonly used and should be your default:
+  it checks for equality within a numerical tolerance:
 
-    ```{r, error = TRUE, eval = FALSE}
+    ```{r, error = TRUE}
     expect_equal(10, 10)
+    expect_equal(10, 10L)
     expect_equal(10, 10 + 1e-7)
     expect_equal(10, 11)
     ```
   
-    If you want to test for exact equivalence, or need to compare a more
-    exotic object like an environment, use `expect_identical()`.
+  If you want to test for exact equivalence or need to compare more exotic
+  objects like environments, use `expect_identical()`.
 
-    ```{r, error = TRUE, eval = FALSE}
+    ```{r, error = TRUE}
     expect_equal(10, 10 + 1e-7)
     expect_identical(10, 10 + 1e-7)
+    expect_equal(2, 2L)
+    expect_identical(2, 2L)
     ```
 
-*   `expect_match()` matches a character vector against a regular expression. 
-    The optional `all` argument controls whether all elements or just one 
-    element needs to match.
-    This is powered by `grepl()` (additional arguments like
-    `ignore.case = FALSE` or `fixed = TRUE` are passed on down).
-
-    ```{r, error = TRUE, eval = FALSE}
+* `expect_match()` matches a character vector against a regular expression. 
+   The optional `all` argument controls whether all elements or just one 
+   element needs to match.
+   This is powered by `grepl()` and additional arguments like
+   `ignore.case = FALSE` or `fixed = TRUE` are passed on down.
+   
+    ```{r, error = TRUE}
     string <- "Testing is fun!"
-
+  
     expect_match(string, "Testing") 
-    
+     
     # Fails, match is case-sensitive
     expect_match(string, "testing")
-
-    # Additional arguments are passed to grepl:
+  
+    # Passes because additional arguments are passed to grepl():
     expect_match(string, "testing", ignore.case = TRUE)
     ```
 
-*   Four variations of `expect_match()` let you check for other types of 
-    result: `expect_output()`, inspects printed output; `expect_message()`,
-    messages; `expect_warning()`, warnings; and `expect_error()` errors.
-    
-    ```{r, error = TRUE, eval = FALSE}
-    a <- list(1:10, letters)
-
-    expect_output(str(a), "List of 2")
-    expect_output(str(a), "int [1:10]", fixed = TRUE)
+* `expect_s3_class()` and `expect_s4_class()` check that an object `inherit()`s
```

Comment:
I think `expect_equal()` + `expect_error()` (and friends) and `expect_snapshot()` are by far the most important and should come first.